Content.Server/Whistle/Systems/WhistleSystem.cs

```diff
@@ -0,0 +1,104 @@
+using Content.Server.Whistle.Components;
+using Content.Shared.Interaction;
+using Content.Shared.Interaction.Events;
+using Content.Shared.Movement.Components;
+using Content.Shared.Timing;
+using Robust.Server.GameObjects;
+
+namespace Content.Server.Whistle;
+
+public class WhistleSystem : EntitySystem
+{
+    [Dependency] private readonly TransformSystem _transform = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly UseDelaySystem _delaySystem = default!;
+    [Dependency] private readonly EntityLookupSystem _entityLookup = default!;
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<WhistleComponent, BeforeRangedInteractEvent>(OnBeforeInteract);
+        SubscribeLocalEvent<WhistleComponent, UseInHandEvent>(OnUseInHand);
+    }
+
+    public void OnBeforeInteract(EntityUid uid, WhistleComponent component, BeforeRangedInteractEvent args)
+    {
+        if (_delaySystem.ActiveDelay(uid))
+            return;
+
+        if(component.exclamatePerson)
+            TryExclamateTarget(uid, args.Target, component);
+
+        _audio.PlayPvs(component.Sound, uid);
+        _delaySystem.BeginDelay(uid);
+
+        args.Handled = true;
+    }
+    private bool TryExclamateTarget(EntityUid uid, EntityUid? target, WhistleComponent component)
+    {
+        if (!HasComp<MobMoverComponent>(target))
+            return false;
+        if (target is null)
+            return false;
+
+        return ExclamateTarget(uid, target.Value, component);
+    }
+    private bool ExclamateTarget(EntityUid uid, EntityUid target, WhistleComponent component)
+    {
+        TransformComponent? targetTransform = null;
+
+        if (!Resolve(target, ref targetTransform))
+            return false;
```

Comment:
As per below you're just using resolves in place of TryGetComponent / passing the component in.